### ISSUE TITLE
[BUG] activatePlugin() raises error

### ISSUE THREAD

**Issue by eltoma-alexkuc** [lucatume/wp-browser#308]
**Environment**
OS: OSX
PHP version: 7.3.9
Installed Codeception version: 2.5.6
Installed wp-browser version: 2.2.23
WordPress version: 2.5.6
Local development environment: docker-compose.yml via Docker Toolbox on OSX
WordPress structure and management: default

**Can you perform the test manually?**
Yes, I can activate plugin via `wp-admin/plugins.php`. No warnings or errors.

**Codeception configuration file**  

```yaml
paths:
    tests: tests
    output: tests/_output
    data: tests/_data
    support: tests/_support
    envs: tests/_envs
actor_suffix: Tester
extensions:
    enabled:
        - Codeception\Extension\RunFailed
    commands:
        - Codeception\Command\GenerateWPUnit
        - Codeception\Command\GenerateWPRestApi
        - Codeception\Command\GenerateWPRestController
        - Codeception\Command\GenerateWPRestPostTypeController
        - Codeception\Command\GenerateWPAjax
        - Codeception\Command\GenerateWPCanonical
        - Codeception\Command\GenerateWPXMLRPC
params:
    - .env.testing
```

**Suite configuration file**

```yaml
actor: AcceptanceTester
modules:
    enabled:
        - WPDb
        - WPBrowser
        - WPFilesystem
        - \Helper\Acceptance
        - WPCLI
        - Asserts
    config:
        WPDb:
            dsn: 'mysql:host=%TEST_SITE_DB_HOST%;dbname=%TEST_SITE_DB_NAME%'
            user: '%TEST_SITE_DB_USER%'
            password: '%TEST_SITE_DB_PASSWORD%'
            dump: 'tests/_data/dump.sql'
            #import the dump before the tests; this means the test site database will be repopulated before the tests.
            populate: true
            # re-import the dump between tests; this means the test site database will be repopulated between the tests.
            cleanup: true
            waitlock: 10
            url: '%TEST_SITE_WP_URL%'
            urlReplacement: true #replace the hardcoded dump URL with the one above
            tablePrefix: '%TEST_SITE_TABLE_PREFIX%'
        WPBrowser:
            url: '%TEST_SITE_WP_URL%'
            adminUsername: '%TEST_SITE_ADMIN_USERNAME%'
            adminPassword: '%TEST_SITE_ADMIN_PASSWORD%'
            adminPath: '%TEST_SITE_WP_ADMIN_PATH%'
        WPFilesystem:
            wpRootFolder: '%WP_ROOT_FOLDER%'
        WPCLI:
            path: '%WP_ROOT_FOLDER%'
            throw: true
            timeout: 60
```

**Describe the bug**
I am unable to activate plugin via `WPBrowser` module, method `activatePlugin()`.

**Output**
```bash
Codeception PHP Testing Framework v2.5.6
Powered by PHPUnit 7.5.15 by Sebastian Bergmann and contributors.
Running with seed: 


  [Connecting To Db] {"config":{"tablePrefix":"wp_","populate":true,"cleanup":true,"reconnect":false,"dump":"tests/_data/dump.sql","populator":null,"urlReplacement":true,"waitlock":10,"dsn":"mysql:host=db:3306;dbname=wordpress","user":"root","password":"root","url":"http://wp.eltoma-global.com"},"options":[]}
  [Db] Connected to default wordpress

Acceptance Tests (1) ---------------------------------------
Modules: WPDb, WPBrowser, WPFilesystem, \Helper\Acceptance, WPCLI, Asserts
------------------------------------------------------------
CsvCest: Import
Signature: CsvCest:Import
Test: tests/acceptance/CsvCest.php:Import
Scenario --
 I am on admin page "plugins.php"
  [Request Headers] []
  [Page] /wp-admin/plugins.php
  [Response] 302
  [Request Cookies] []
  [Response Headers] {"Date":["Sun, 15 Sep 2019 17:50:34 GMT"],"Server":["Apache/2.4.38 (Debian)"],"X-Powered-By":["PHP/7.3.9"],"Expires":["Wed, 11 Jan 1984 05:00:00 GMT"],"Cache-Control":["no-cache, must-revalidate, max-age=0"],"X-Redirect-By":["WordPress"],"Location":["http://wp.eltoma-global.com/wp-login.php?redirect_to=http%3A%2F%2Fwp.eltoma-global.com%2Fwp-admin%2Fplugins.php&reauth=1"],"Content-Length":["0"],"Content-Type":["text/html; charset=UTF-8"]}
  [Redirecting to] http://wp.eltoma-global.com/wp-login.php?redirect_to=http%3A%2F%2Fwp.eltoma-global.com%2Fwp-admin%2Fplugins.php&reauth=1
  [Page] http://wp.eltoma-global.com/wp-login.php?redirect_to=http%3A%2F%2Fwp.eltoma-global.com%2Fwp-admin%2Fplugins.php&reauth=1
  [Response] 200
  [Request Cookies] []
  [Response Headers] {"Date":["Sun, 15 Sep 2019 17:50:35 GMT"],"Server":["Apache/2.4.38 (Debian)"],"X-Powered-By":["PHP/7.3.9"],"Expires":["Wed, 11 Jan 1984 05:00:00 GMT"],"Cache-Control":["no-cache, must-revalidate, max-age=0"],"Set-Cookie":["wordpress_test_cookie=WP+Cookie+check; path=/","wordpress_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/wp-admin","wordpress_sec_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/wp-admin","wordpress_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/wp-content/plugins","wordpress_sec_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/wp-content/plugins","wordpress_logged_in_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpress_logged_in_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wp-settings-0=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wp-settings-time-0=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpress_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpress_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpress_sec_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpress_sec_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpressuser_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpresspass_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpressuser_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wordpresspass_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/","wp-postpass_a0fd6ad7d4c03dfafce4e3e367ce11fe=+; expires=Sat, 15-Sep-2018 17:50:35 GMT; Max-Age=0; path=/"],"X-Frame-Options":["SAMEORIGIN"],"Vary":["Accept-Encoding"],"Content-Length":["2941"],"Content-Type":["text/html; charset=UTF-8"]}
 I cli "plugin install hello-dolly --force"
  [WPCLI command] ["plugin","install","hello-dolly","--force"]
  [WPCLI command with configuration options] ["plugin","install","hello-dolly","--force"]
  [WPCLI command with environment] []
  [WPCLI output] Installing Hello Dolly (1.7.2)
  Downloading installation package from https://downloads.wordpress.org/plugin/hello-dolly.1.7.2.zip...
  Using cached file '/home/admin/.wp-cli/cache/plugin/hello-dolly-1.7.2.zip'...
  Unpacking the package...
  Installing the plugin...
  Removing the old version of the plugin...
  Plugin updated successfully.
  Success: Installed 1 of 1 plugins.
  
  [WPCLI  status] 0
 I cli "plugin list"
  [WPCLI command] ["plugin","list"]
  [WPCLI command with configuration options] ["plugin","list"]
  [WPCLI command with environment] []
  [WPCLI output] name	status	update	version
  wp_plugin	inactive	none	0.1
  hello-dolly	inactive	none	1.7.2
  
  [WPCLI  status] 0
 I activate plugin "hello-dolly"
 FAIL 

------------------------------------------------------------
19x DEPRECATION: Calling the "Symfony\Component\EventDispatcher\EventDispatcherInterface::dispatch()" method with the event name as the first argument is deprecated since Symfony 4.3, pass it as the second argument and provide the event object as the first argument instead. /app/wp-content/plugins/wp_plugin/vendor/symfony/event-dispatcher/EventDispatcher.php:58
4x DEPRECATION: The "Symfony\Component\BrowserKit\Response::getStatus()" method is deprecated since Symfony 4.3, use getStatusCode() instead. /app/wp-content/plugins/wp_plugin/vendor/symfony/browser-kit/Response.php:100
2x DEPRECATION: Passing a command as string when creating a "Symfony\Component\Process\Process" instance is deprecated since Symfony 4.2, pass it as an array of its arguments instead, or use the "Process::fromShellCommandline()" constructor if you need features provided by the shell. /app/wp-content/plugins/wp_plugin/vendor/symfony/process/Process.php:147


Time: 7.8 seconds, Memory: 6.00 MB

There was 1 failure:

---------
1) CsvCest: Import
 Test  tests/acceptance/CsvCest.php:Import
 Step  Activate plugin "hello-dolly"
 Fail  Form field by Label or CSS element with '//*[@data-slug="hello-dolly"]/th/input' was not found.

Scenario Steps:

 4. $I->activatePlugin("hello-dolly") at tests/acceptance/CsvCest.php:28
 3. $I->cli("plugin list") at tests/acceptance/CsvCest.php:27
 2. $I->cli("plugin install hello-dolly --force") at tests/acceptance/CsvCest.php:26
 1. $I->amOnAdminPage("plugins.php") at tests/acceptance/CsvCest.php:25


FAILURES!
Tests: 1, Assertions: 0, Failures: 1.
  [Db] Disconnected from default
```

**To Reproduce**
```php
<?php

class CsvCest
{
    public function _before(AcceptanceTester $I) {}

    public function _after(AcceptanceTester $I) {}

    public function Import(AcceptanceTester $I)
    {
        $I->cli("plugin install hello-dolly --force");
        $I->amOnAdminPage("plugins.php");
        $I->cli("plugin list");
        $I->activatePlugin("hello-dolly");
        $I->seePluginActivated("hello-dolly");
        $I->cli("plugin list");
    }
}
```

**Expected behavior**
Plugin should be activated.

**Comment by lucatume** [comment]
@eltoma-alexkuc from the test code you posted I see you're not logging in the site using `loginAsAdmin` or `loginAs` method.
Add that in the test case `_before` method or at the start of your test methods and let me know if that solves the issue.

You can see this from the debug output:
```
I am on admin page "plugins.php"
  [Request Headers] []
  [Page] /wp-admin/plugins.php
  [Response] 302
  [Request Cookies] []
  [Response Headers] {"Date":["Sun, 15 Sep 2019 17:50:34 GMT"],"Server":["Apache/2.4.38 (Debian)"],"X-Powered-By":["PHP/7.3.9"],"Expires":["Wed, 11 Jan 1984 05:00:00 GMT"],"Cache-Control":["no-cache, must-revalidate, max-age=0"],"X-Redirect-By":["WordPress"],"Location":["http://wp.eltoma-global.com/wp-login.php?redirect_to=http%3A%2F%2Fwp.eltoma-global.com%2Fwp-admin%2Fplugins.php&reauth=1"],"Content-Length":["0"],"Content-Type":["text/html; charset=UTF-8"]}
  [Redirecting to] http://wp.eltoma-global.com/wp-login.php?redirect_to=http%3A%2F%2Fwp.eltoma-global.com%2Fwp-admin%2Fplugins.php&reauth=1
  [Page] http://wp.eltoma-global.com/wp-login.php?redirect_to=http%3A%2F%2Fwp.eltoma-global.com%2Fwp-admin%2Fplugins.php&reauth=1
  [Response] 200
```

**Comment by mityung** [comment]
@lucatume I analysed it further and found that it happen only in a given scenario
1. We need to set some flags in our Cest(wpbrowser test) files and use those flags in our plugin code.
2. We are using wp cli to set those flags in Cest files(wp cli module enabled and working fine)
`$I->cli("config set DUMMYFLAG 1 --raw");` 
3. We are using SemaphoreCI to run tests.

So, idea is to set these flags before activating plugin in Cest file, then in plugin activation hook, check if those flags are defined(using defined()), and perform some work conditionally for each Cest.

**Problem:**
When using SemaphoreCI(working fine on local setup), If we use activatePlugin() to activate the plugin, these flag shows undefined(even if they exists in wp-config.php)
if we use wp cli to activate plugin, these flags working correctly
`$I->cli('plugin activate myplugin');`

These Constants are correctly written in wp-config.php, i tested by printing content of wp-config.php file after plugin activation inside Cest file.

Our Conclusion:
wpbrowser test not loading wp-config.php before calling activatePlugin(), while cli module loads wp-config.php again, before activating plugin.

It can be proved by activating another random plugin using wp cli module and then activating our plugin using activatePlugin() method, it's successful, because while activating random plugin, wp cli reloaded wp-config.php.

Not working:
`$I->cli("config set DUMMYFLAG 1 --raw");
 $I->amOnAdminPage("plugins.php");
 $I->activatePlugin('csv-importer-custom');` 

Working:
`$I->cli("config set DUMMYFLAG 1 --raw");
 $I->cli('plugin activate myplugin');` 

Working:
`$I->cli("config set DUMMYFLAG 1 --raw");
 $I->cli('plugin activate hello-dolly');
 $I->amOnAdminPage("plugins.php");
 $I->activatePlugin('csv-importer-custom');` 

_**NOTE: All 3 cases above, working fine on local system, Problem is only when run tests in SemaphoreCI.**_

After settings flags in config in Cest file, I activated plugin hello-dolly using wp cli module and then used activatePlugin() to activate csv import plugin and it's successful.

Is it because activatePlugin() not force wordpress to reload wp-config.php and wp cli activate plugin force it?

**Comment by lucatume** [comment]
@mityung could you please paste the Cest code you're using?

Per https://github.com/lucatume/wp-browser/issues/308#issuecomment-531668128: I do not see where you're using `loginAsAdmin` and can see, in the output, where the `WPBrowser` module tries to redirect to the login screen.

Before delving into specific CI/env issues this should work locally, does it?

**Comment by mityung** [comment]
@lucatume as i written in above **NOTE** that it's working fine on local system.
here is my complete Cest code

`class CsvCest
 {
    public function _before(AcceptanceTester $I)
    {
        $I->loginAsAdmin();
        $I->amOnAdminPage("themes.php");
        $I->useTheme('TESTTHEME');
        $I->cli("config set TESTFLAG 1 --raw --allow-root");
    }
    public function _after(AcceptanceTester $I)
    {
        $I->cli("config delete TESTFLAG --allow-root");
    }
    public function Import(AcceptanceTester $I)
    {
        $I->amOnAdminPage("plugins.php");
        $I->activatePlugin('csv-importer-custom'); //Not working
        // $I->cli('plugin activate wp_plugin --user=1');  //Working
    }
  }
`

**Comment by lucatume** [comment]
@mityung thanks for providing the example code; this helps me a lot to understand your problem.

> wpbrowser test not loading wp-config.php before calling activatePlugin()

No, it does not load any WordPress files. 
WPBrowser will browse on the plugins page and click to activate the plugin as a user would. The plugin process activation is handled via the webserver.

I have to look into this to provide a correct answer: are you using containers on the CI system? At times there might be a 2-3 second delay between writing a file on the host (as wp-cli does) and it being synced in the guest container (that running the webserver).

Since wp-cli executes in a different process and writes to the file, that is immediately synced. A hint that this might be the case is that it works on the faster, and dedicated local setup.

I've solved the issue, in the past, by adding a wait time of 3 seconds after a modification to the underlying files when in the container environment.

With [Codeception 3.0 you could try the `retry` prefix](https://codeception.com/docs/03-AcceptanceTests#retry) for the `activatePlugin` call.

I'm not saying this is the solution, but it might be, at least, a 

On a side note: you can use `allow-root: true` in the WPCLI configuration file to have that appended for you by the module.

**Comment by mityung** [comment]
@lucatume Thanks for reply,

> I have to look into this to provide a correct answer: are you using containers on the CI system?

Yes, we are using containers

I had same doubt and tried to print contents of wp-config.php file just before and after activating plugin using `activatePlugin`
```php
    public function Import(AcceptanceTester $I)      
    {      
        $I->amOnAdminPage("plugins.php");   
        $configText1 = file_get_contents('/PATH_TO_WP_ROOT/wp-config.php');
        codecept_debug('==============');
        codecept_debug($configText1);
        codecept_debug('==============');   
        $I->activatePlugin('myplugin'); //Not working      
        $configText2 = file_get_contents('/PATH_TO_WP_ROOT/wp-config.php');
        codecept_debug('==============');
        codecept_debug($configText2);
        codecept_debug('==============');
        // $I->cli('plugin activate myplugin --user=1');  //Working      
    }  

```
And it showing flags are there in content of wp-config.php file on both prints.
I'm not sure If file is not synced with wpbrowser environment, can it print those flags in content of wp-config.php?

**Comment by lucatume** [comment]
I most definitely have to look into this.
Thanks for the debugging information. I will update with further information as I have it.

**Comment by eltoma-alexkuc** [comment]
### TL;DR

@lucatume 

> I have to look into this to provide a correct answer: are you using containers on the CI system? At times there might be a 2-3 second delay between writing a file on the host (as wp-cli does) and it being synced in the guest container (that running the webserver).

This solved it! Did 10 runs on CI and none of them failed. I will clarify why there was a failure.

### Detailed analysis

We use custom WordPress plugin which imports data from CSV but for performance reasons we do _selective_ import using so-called flags (global constants).

So the interesting thing about this quirky bug is that a) it occurs only in CI but only 30%-50% of the time i.e. not always b) it doesn't happen on local machines (verified via 2 entirely different OS with entirely different hardware).

So here's a breakdown of what happens:

- using [file_get_contents](https://www.php.net/manual/en/function.file-get-contents.php) we can see that the code for flags (`define(...);`) is there

- **but!** using [get_defined_constants](https://www.php.net/manual/en/function.get-defined-constants.php), we can see that when CI test fails with this random error, the expect flags (globals) are **not** there

- failure (`[PHPUnit\Framework\Exception] Undefined offset: 0  `) was actually caused by this line (2nd line):

```php
$id_arr1 = $I->grabAllFromDatabase('wp_posts', 'id', array('post_title' => 'Company Registration and Formation Process in Collaboration with Eltoma Global'));
$id1 = $id_arr1[0]["id"];
```

- to clarify, as flags (global constants) did not exist in the PHP memory, csv plugin wasn't importing anything at all so no posts were seen in db

### Summary

To summarise: while custom settings (global constants) existed physically in the file, they did **not** exist in PHP memory as confirmed by [file_get_contents](https://www.php.net/manual/en/function.file-get-contents.php) and [get_defined_constants](https://www.php.net/manual/en/function.get-defined-constants.php).

### Speculations

This suggests there is an implicit race condition which is exacerbated by high-performance systems like CIs where speed of execution is top priority.

After analyzing [WPCLI.php](https://github.com/lucatume/wp-browser/blob/639523f96211fc525dd7bdc6bfbd2ee14591eae6/src/Codeception/Module/WPCLI.php), I can see that `wp-cli` is invoked via `executeWpCliCommand` which in turn calls [mustRun](https://symfony.com/doc/current/components/process.html) which is similar to [run](https://symfony.com/doc/current/components/process.html). Both are blocking commands i.e. waits for the shell to finish before proceeding.

From this I can speculate that either, it is a file system quirk caused by Docker or PHP code is not waiting for shell to complete execution of `wp-cli` though myself I couldn't find which could runs `wp-cli` asynchronously.

### Solution

- either wait for 2-3 seconds as suggested by @lucatume via [sleep](https://www.php.net/manual/en/function.sleep.php)

- or use `$I->cli(...)` which does not experience this quirky condition/bug

**Comment by lucatume** [comment]
@eltoma-alexkuc thanks for getting back with such a wealth of information.
It's good to know you, and your team found why this happens and how to get around it.

In many suites I work around this issue in a similar way, using variations of this module:

```php
<?php

namespace Helper;

use Codeception\Module;

class Container extends Module
{
    public function waitForFile($file = null)
    {
        if (!(getenv('CI') || getenv('CONTAINER'))) {
            codecept_debug('Not in CI/Container context: not waiting');
            return;
        }

        $waitFor = $this->config['wait'] ?: 3;

        if ($file === null) {
            // Just wait some arbitrary time if we're in CI/Container context.
            codecept_debug('In CI/Container context: sleeping for ' . $waitFor);
            sleep($waitFor);
            return;
        }

        codecept_debug('In CI/Container context: waiting for file ' . $file . ' for ' . $waitFor);

        $time = 0;
        while ($time < $waitFor) {
            if (file_exists($file)) {
                return;
            }

            $time += microtime(true);
        }

        throw new \RuntimeException('Waited for ' . $waitFor . ' but file still does not exist.');
    }
}
```

I put this in `tests/_support/Helper/ContainerContext.php` and enable the module on any suite that might be affected by it.
You can configure the module using the `wait` parameter (or it will default to 3 seconds):

```yaml
actor: CommandTester
modules:
    enabled:
        - \Helper\Command
        - Asserts
        - \Helper\Container
    config:
        \Helper\Container:
            wait: 5
```

If you think there's no more work to do here, please close the issue.

**Comment by eltoma-alexkuc** [comment]
@lucatume This is excellent suggestion, thank you very much! I believe at this point the issue can be considered resolved unless you have any other questions?
